src/quota_operation_aggregator.cc

```diff
@@ -0,0 +1,128 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "src/quota_operation_aggregator.h"
+
+#include <iostream>
+
+#include <google/protobuf/text_format.h>
+
+#include "src/money_utils.h"
+#include "src/signature.h"
+#include "utils/distribution_helper.h"
+#include "utils/stl_util.h"
+
+using std::string;
+using ::google::protobuf::Timestamp;
+using google::api::MetricDescriptor;
+using google::api::servicecontrol::v1::MetricValue;
+using google::api::servicecontrol::v1::MetricValueSet;
+using google::api::servicecontrol::v1::QuotaOperation;
+
+namespace google {
+namespace service_control_client {
+
+namespace {
+
+// Returns whether timestamp a is before b or not.
+bool TimestampBefore(const Timestamp& a, const Timestamp& b) {
+  return a.seconds() < b.seconds() ||
+         (a.seconds() == b.seconds() && a.nanos() < b.nanos());
+}
+
+// Merges two metric values, with metric kind being Delta.
+//
+// For INT64, values will be added together,
+// except no change when the bucket options does not match.
+void MergeDeltaMetricValue(const MetricValue& from, MetricValue* to) {
+  if (to->value_case() != from.value_case()) {
+    GOOGLE_LOG(WARNING) << "Metric values are not compatible: "
+                        << from.DebugString() << ", " << to->DebugString();
+    return;
+  }
+
+  if (from.has_start_time()) {
+    if (!to->has_start_time() ||
+        TimestampBefore(from.start_time(), to->start_time())) {
+      *(to->mutable_start_time()) = from.start_time();
+    }
+  }
+
+  if (from.has_end_time()) {
+    if (!to->has_end_time() ||
+        TimestampBefore(to->end_time(), from.end_time())) {
+      *(to->mutable_end_time()) = from.end_time();
+    }
+  }
+
+  switch (to->value_case()) {
+    case MetricValue::kInt64Value:
+      to->set_int64_value(to->int64_value() + from.int64_value());
+      break;
+    default:
+      GOOGLE_LOG(WARNING) << "Unknown metric kind for: " << to->DebugString();
+      break;
+  }
+}
+
+}  // namespace anonymous
+
+QuotaOperationAggregator::QuotaOperationAggregator(
+    const ::google::api::servicecontrol::v1::QuotaOperation& operation)
+    : operation_(operation), is_aggregated_(false) {
+  MergeOperation(operation);
+}
+
+void QuotaOperationAggregator::MergeOperation(const QuotaOperation& operation) {
+  for (const auto& metric_value_set : operation.quota_metrics()) {
+    // Intentionally use the side effect of [] to add missing keys.
+    std::unordered_map<string, MetricValue>& metric_values =
+        metric_value_sets_[metric_value_set.metric_name()];
+
+    for (const auto& metric_value : metric_value_set.metric_values()) {
+      if(metric_value.int64_value() > 0) {
+        is_aggregated_ = true;
+      }
+
+      string signature = GenerateReportMetricValueSignature(metric_value);
```

Comment:
This is not working.   if we have multiple values, they don't have any labels(),  then signature will be the same.  

We only have one value,  not need to create map {string -> map {string, value}}
just map {string -> value}

and assert metirc_values_size() is 1,  and always use metric_values(0)